sdks/python/apache_beam/runners/direct/evaluation_context.py

```diff
@@ -69,67 +69,109 @@ class _SideInputsContainer(object):
-      view = self._views[side_input]
-      assert not view.has_result
-      assert view.value is None
-      assert view.callable_queue is not None
-      view.value = self._pvalue_to_value(side_input, view.elements)
-      view.elements = None
-      result = tuple(view.callable_queue)
-      for task in result:
-        task.blocked = False
-      view.callable_queue = None
-      view.has_result = True
-      return result
-
-  def update_watermarks_for_transform(self, ptransform, watermark):
-    # Collect tasks that get unblocked as the workflow progresses.
-    unblocked_tasks = []
-    for view in self._transform_to_views[ptransform]:
-      unblocked_tasks.extend(self._update_watermarks_for_view(view, watermark))
-    return unblocked_tasks
+  def update_watermarks_for_transform_and_unblock_tasks(self,
+                                                        ptransform,
+                                                        watermark):
+    """Updates _SideInputsContainer after a watermark update and unbloks tasks.
+
+    It traverses the list of side inputs per PTransform and calls
+    _update_watermarks_for_side_input_and_unblock_tasks to unblock tasks.
+
+    Args:
+      ptransform: Value of a PTransform.
+      watermark: Value of the watermark after an update for a PTransform.
 
-  def _update_watermarks_for_view(self, view, watermark):
+    Returns:
+      Tasks that get unblocked as a result of the watermark advancing.
+    """
     unblocked_tasks = []
-    if watermark.input_watermark == WatermarkManager.WATERMARK_POS_INF:
-      unblocked_tasks = self.finalize_value_and_get_tasks(view)
+    for side in self._transform_to_side_inputs[ptransform]:
+      unblocked_tasks.extend(
+          self._update_watermarks_for_side_input_and_unblock_tasks(
+              side, watermark))
     return unblocked_tasks
 
-  def _pvalue_to_value(self, view, values):
-    """Given a side input view, returns the associated value in requested form.
+  def _update_watermarks_for_side_input_and_unblock_tasks(self,
+                                                          side_input,
+                                                          watermark):
+    """Helps update _SideInputsContainer after a watermark update.
+
+    For each view of the side input, it updates the value of the watermark
+    recorded when the watermark moved and unblocks tasks accordingly.
+
+    Args:
+      side_input: (_UnpickledSideInput) value.
+      watermark: Value of the watermark after an update for a PTransform.
+
+    Returns:
+      Tasks that get unblocked as a result of the watermark advancing.
+    """
+    with self._lock:
+      unblocked_tasks = []
+      view = self._views[side_input]
+      view.watermark = watermark
+
+      # Unblock and finalize tasks
+      view = self._views[side_input]
+      tasks_to_remove = []
+      for task, block_until in view.blocked_tasks:
+        if watermark.input_watermark >= block_until and view.elements:
```

Comment:
<!--new_thread; commit:0d82258b7674656d3a6f86748f47d23de8cf512f; resolved:0-->
Why do we need the `and view.elements`?  It can happen that the view has no elements, right?  In such a case, do we want to keep the task in "limbo"?

Reply:
Good point (it was an old debugging condition). Done!